### ISSUE TITLE
Inconsistent `sum`/`dot`/`norm`$^2$ behavior

### ISSUE THREAD

**Issue by melnikovsky** [pytorch/pytorch#151761]
### 🐛 Describe the bug

Summation of huge `float32` arrays is admittedly a sensitive subject, but different routines use inconsistent (and seemingly undocumented?) approaches. Particularly, `torch.sum` is the most precise, while `linalg.norm` on 10 CPU cores is as slow but has inferior accuracy. Would it be possible to normalize these somehow? How do I get consistent results even in future `pytorch` versions?
Below is the output for three functions, the right answer is `1e9`

```
         1  thread:
torch.linalg.norm(x)**2=tensor(6.5533e+08),   timeit: [1.0846478752791882, 1.0839017871767282, 1.0842528380453587]
torch.dot(x,x)=tensor(9.7329e+08),            timeit: [1.0753544569015503, 1.075887642800808, 1.075775207951665]
(x*x).sum()=tensor(1.0000e+09),               timeit: [4.653062522411346, 4.647735759615898, 4.65124611929059]

         10 threads:
torch.linalg.norm(x)**2=tensor(6.5533e+08),   timeit: [1.0826967414468527, 1.0804776344448328, 1.078405149281025]
torch.dot(x,x)=tensor(9.9902e+08),            timeit: [0.2012637760490179, 0.2010939735919237, 0.20179643481969833]
(x*x).sum()=tensor(1.0000e+09),               timeit: [1.0688033681362867, 1.0729365721344948, 1.0708447061479092]

```

The code itself:
```
import torch
import timeit

def play(x):
   print(f'{torch.linalg.norm(x)**2=},   timeit:', timeit.repeat('torch.linalg.norm(x)**2', number=4, repeat=3, globals=globals() ))
   print(f'{torch.dot(x,x)=},   timeit:', timeit.repeat('torch.dot(x,x)', number=4, repeat=3, globals=globals()))
   print(f'{(x*x).sum()=},   timeit:', timeit.repeat('(x*x).sum()', number=4, repeat=3, globals=globals()))

x=torch.randn(1_000_000_000, dtype=torch.float32, device='cpu')

torch.set_num_threads(1)
print('\t 1  thread:')
play(x)

torch.set_num_threads(10)
print('\n\t 10 threads:')
play(x)
```

### Versions

Collecting environment information...
PyTorch version: 2.6.0
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Red Hat Enterprise Linux release 8.10 (Ootpa) (x86_64)
GCC version: (GCC) 12.2.0
Clang version: Could not collect
CMake version: version 3.26.5
Libc version: glibc-2.28

Python version: 3.13.3 | packaged by conda-forge | (main, Apr 14 2025, 20:44:03) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-4.18.0-553.27.1.el8_10.x86_64-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A40
Nvidia driver version: 560.35.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              24
On-line CPU(s) list: 0-23
Thread(s) per core:  1
Core(s) per socket:  12
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               106
Model name:          Intel(R) Xeon(R) Gold 5317 CPU @ 3.00GHz
Stepping:            6
CPU MHz:             3000.000
CPU max MHz:         3600.0000
CPU min MHz:         800.0000
BogoMIPS:            6000.00
L1d cache:           48K
L1i cache:           32K
L2 cache:            1280K
L3 cache:            18432K
NUMA node0 CPU(s):   0,2,4,6,8,10,12,14,16,18,20,22
NUMA node1 CPU(s):   1,3,5,7,9,11,13,15,17,19,21,23
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==2.2.5
[pip3] optree==0.14.1
[pip3] torch==2.6.0
[pip3] triton==3.2.0+git576374f8
[conda] cuda-cudart               12.8.90              h5888daf_1    conda-forge
[conda] cuda-cudart_linux-64      12.8.90              h3f2d84a_1    conda-forge
[conda] cuda-cupti                12.8.90              h5888daf_1    conda-forge
[conda] cuda-nvrtc                12.8.93              h5888daf_1    conda-forge
[conda] cuda-nvtx                 12.8.90              h5888daf_1    conda-forge
[conda] cudnn                     9.8.0.87             h81d5506_1    conda-forge
[conda] libblas                   3.9.0           31_hfdb39a5_mkl    conda-forge
[conda] libcblas                  3.9.0           31_h372d94f_mkl    conda-forge
[conda] libcublas                 12.8.4.1             h9ab20c4_1    conda-forge
[conda] libcufft                  11.3.3.83            h5888daf_1    conda-forge
[conda] libcurand                 10.3.9.90            h9ab20c4_1    conda-forge
[conda] libcusolver               11.7.3.90            h9ab20c4_1    conda-forge
[conda] libcusparse               12.5.8.93            h5888daf_1    conda-forge
[conda] liblapack                 3.9.0           31_hc41d3b0_mkl    conda-forge
[conda] libmagma                  2.9.0                h19665d7_1    conda-forge
[conda] libnvjitlink              12.8.93              h5888daf_1    conda-forge
[conda] libtorch                  2.6.0           cuda126_mkl_h99b69db_304    conda-forge
[conda] mkl                       2024.2.2            ha957f24_16    conda-forge
[conda] nccl                      2.26.2.1             ha44e49d_1    conda-forge
[conda] numpy                     2.2.5           py313h17eae1a_0    conda-forge
[conda] optree                    0.14.1          py313hdb19cb5_0  
[conda] pytorch                   2.6.0           cuda126_mkl_py313_he20fe19_304    conda-forge
[conda] pytorch-gpu               2.6.0           cuda126_mkl_ha999a5f_304    conda-forge
[conda] triton                    3.2.0           cuda126py313h46f6bd1_1    conda-forge

**Comment by malfet** [comment]
float32 has 23-bits of precision, 30-bit are needed to encode 1_000_000_000, so indeed there are expected to be some numerical inaccuracy when accumulating using float32 datatype. Although indeed linalg.norm looks somewhat wrong, but may be within the expect accuracy

**Comment by lezcano** [comment]
sum implements kahan algorithm to minimise the error. Probably our norm implementation doesn't.

There are no hard promises for precision between versions. If we regress we will fix it, but we can't say anything else beyond what's discussed at https://pytorch.org/docs/stable/notes/numerical_accuracy.html

**Comment by melnikovsky** [comment]
Well, if more than 10-fold accuracy inconsistency between `torch.linalg.norm(x)**2` and  `torch.dot(x,x)` does not qualify as regression... Maybe a 35% result discrepancy between CPU and GPU versions of the same code could be addressed:

```
torch.linalg.norm(x1)**2=tensor(1.0000e+09, device='cuda:0')
torch.linalg.norm(x2)**2=tensor(6.5530e+08, device='cpu')
```
Thank you!

**Comment by ngimel** [comment]
>sum implements kahan algorithm to minimise the error. Probably our norm implementation doesn't. 

sum does better than kahan - it uses cascade summation #39516 and indeed norm doesn't. And yes, we'll accept PRs updating norm to use the same implementation

**Comment by ngimel** [comment]
see also #65815

**Comment by nobody1232242** [comment]
This is a floating-point error, likely caused by differences in the order of operations. I suggest closing this issue.
